### PR TITLE
Work with latest mypy

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3734

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -155,7 +155,11 @@ basepython = python3
 skip_install = True
 deps =
     mypy
-    git+https://github.com/Shoobx/mypy-zope
+    mypy-zope
+    types-mock
```
Comment (itamarst) [inline tox.ini:159]: For explanation of "types-*" packages see http://mypy-lang.blogspot.com/2021/06/mypy-0900-released.html

File: tox.ini [hunk 2]
```diff
@@ -155,7 +155,11 @@ basepython = python3
 skip_install = True
 deps =
     mypy
-    git+https://github.com/Shoobx/mypy-zope
+    mypy-zope
```
Comment (itamarst) [inline tox.ini:158]: The mypy-zope change is unrelated, just figured that if I'm changing this I'd check if mypy-zope had done a new release, and turns out they had.

**Review by sajith — APPROVED** [review]
Looks good to me.  I have just one (possibly goofy) question.

I did a `source .tox/typechecks/bin/activate`, and then ran `mypy src` directly.  It printed a bunch of error messages:
  
```
src/allmydata/test/test_consumer.py:25: error: 'Producer' is missing following 'twisted.internet.interfaces.IProducer' interface members: stopProducing.
src/allmydata/test/test_consumer.py:25: error: 'Producer' is missing following 'IPushProducer' interface members: pauseProducing.
src/allmydata/util/deferredutil.py:217: error: "Deferred" expects no type arguments, but 1 given
src/allmydata/util/deferredutil.py:217: error: The return type of a generator function should be "Generator" or one of its supertypes
src/allmydata/util/deferredutil.py:218: error: "Deferred" expects no type arguments, but 1 given
src/allmydata/storage_client.py:150: error: 'StorageFarmBroker' is missing following 'IStorageBroker' interface members: get_all_connections, get_all_connectors, get_all_peerids, get_all_connections_for, get_permuted_peers.
src/allmydata/introducer/client.py:39: error: 'IntroducerClient' is missing following 'RIIntroducerSubscriberClient_v2' interface members: __remote_name__, announce_v2.
src/allmydata/storage/server.py:54: error: 'StorageServer' is missing following 'RIStorageServer' interface members: __remote_name__, get_version, allocate_buckets, add_lease, renew_lease, get_buckets, slot_readv, advise_corrupt_share.
src/allmydata/storage/server.py:55: error: Incompatible types in assignment (expression has type "str", base class "Service" defined the type as "None")
src/allmydata/testing/web.py:252: error: '_SynchronousProducer' is missing following 'IBodyProducer' interface members: stopProducing.
src/allmydata/testing/web.py:252: error: '_SynchronousProducer' is missing following 'twisted.internet.interfaces.IPushProducer' interface members: pauseProducing, resumeProducing.
src/allmydata/immutable/upload.py:1857: error: Incompatible types in assignment (expression has type "str", base class "Service" defined the type as "None")
src/allmydata/control.py:69: error: 'ControlServer' is missing following 'RIControlClient' interface members: wait_for_client_connections, upload_random_data_from_file, download_to_tempfile_and_delete, get_memory_usage, speed_test, measure_peer_response_time.
src/allmydata/web/common.py:735: error: "IRequest" has no attribute "fields"
src/allmydata/web/common.py:736: error: "IRequest" has no attribute "fields"
src/allmydata/frontends/sftpd.py:1918: error: 'FakeTransport' is missing following 'ITransport' interface members: getPeer, getHost.
src/allmydata/frontends/sftpd.py:1993: error: Signature of "Dispatcher" incompatible with "requestAvatar" of supertype "IRealm"
src/allmydata/frontends/sftpd.py:2001: error: Incompatible types in assignment (expression has type "str", base class "Service" defined the type as "None")
src/allmydata/web/operations.py:47: error: Incompatible types in assignment (expression has type "str", base class "Service" defined the type as "None")
src/allmydata/webish.py:234: error: Incompatible types in assignment (expression has type "str", base class "Service" defined the type as "None")
src/allmydata/introducer/server.py:162: error: 'IntroducerService' is missing following 'RIIntroducerPublisherAndSubscriberService_v2' interface members: __remote_name__, get_version, publish_v2, subscribe_v2.
src/allmydata/introducer/server.py:163: error: Incompatible types in assignment (expression has type "str", base class "Service" defined the type as "None")
Found 22 errors in 13 files (checked 275 source files)
```

Why is running `tox -r typechecks` different from running `mypy src`?

I don't know if some missing interface members and some type violations inherited from libraries are worth worrying about, but I am pretty sure they're not in the scope of this PR.   Please merge anyway.

**Comment by itamarst** [comment]
The differences might be the Twisted version? Specifically https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3622
